docs/ENVIRONMENTALS.md
## 📝 Mangatsu Server - Configuration
Usable options inside the **.env** or **docker-compose.yml**:

_~~Struck out~~ values have no effect yet._

- **MTSU_ENV**=production
    - Environment: production, development
- **MTSU_LOG_LEVEL**=info
    - Log level: debug, info, warn, error
- **MTSU_INITIAL_ADMIN_NAME**=admin
- **MTSU_INITIAL_ADMIN_PW**=admin321
    - Credentials for the initial admin user. Recommended to change.
- **MTSU_DOMAIN**: example.org.
    - For example, if the address for the server is "api.example.com", and for the frontend "read.example.com", the value here should be "example.com" for the cookies to work properly between subdomains.
    - https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#define_where_cookies_are_sent
- **MTSU_STRICT_ACAO**: 'true'
    - When true, the server only allows authenticated connections from the MTSU_DOMAIN and its subdomains (eg .*.example.org). When false, connections from every origin is allowed.
    - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
- **MTSU_HOSTNAME**=localhost
- **MTSU_PORT**=5050
    - Hostname and port for the server. Use **mtsuserver** as the hostname if using Docker Compose.
- **MTSU_BASE_PATHS**=freeform1;/home/user/doujinshi;;structured2;/home/user/manga
    - Paths to the archive directories. Relative or absolute paths are accepted.
    - First specify the type of the directory and a numerical ID (e.g. freeform1 or structured2) and then the path separated by a semicolon: `;`.
    - Multiple paths can be separated by a double-semicolon: `;;`.
    - Format: `<freeform|structured><ID>;<INTERNAL_PATH>;;<freeform|structured><ID>;<INTERNAL_PATH>`...
    - If using Docker Compose, make sure that the paths match the containerpaths in the volumes section.
- **MTSU_DATA_PATH**=../data
    - Location of the data dir which includes the SQLite db and the cache for gallery images and thumbnails. Relative or absolute paths are accepted.
    - Doesn't need changing if using Docker Compose.
- **MTSU_DISABLE_CACHE_SERVER**=false
    - Set true to disable the internal cache server (serves media files and thumbnails). Useful if one wants to use the web server such as NGINX to serve the files.
- **MTSU_CACHE_TTL**=336h
    - Cache time to live (for example `336h` (2 weeks), `8h30m`). If a gallery is not viewed for this time, it will be purged from the cache.
- ~~**MTSU_CACHE_SIZE**~~=10000
    - Max size of the cache where galleries are extracted from the library in MB. Can overflow a bit especially if set too low.
- **MTSU_DB_NAME**=mangatsu
    - Name of the SQLite database file
- ~~**MTSU_DB**~~=sqlite
    - Database type: `sqlite`, `postgres`, `mysql` or `mariadb`
- ~~**MTSU_DB_HOST**~~=localhost
    - Hostname of the database server.
- ~~**MTSU_DB_PORT**~~=5432
    - Usually 5432 for PostgreSQL and 3306 for MySQL and MariaDB.
- ~~**MTSU_DB_USER**~~=mtsu-user
- ~~**MTSU_DB_PASSWORD**~~=s3cr3t
- **MTSU_VISIBILITY**=public
    - **public**: anyone can access the collection and its galleries.
    - **restricted**: users need a global passphrase to access collection and its galleries.
    - **private**: only logged-in users can access the collection and its galleries.
    - In all modes, user accounts are supported and have more privileges than anonymous users (e.g. favorite galleries).
- **MTSU_RESTRICTED_PASSPHRASE**=secretpassword
    - Passphrase to access the collection and its galleries.
    - Only used when **VISIBILITY** is set to **restricted**.
- **MTSU_REGISTRATIONS**=false
    - Whether to allow user registrations. If set to false, only admins can create new users.
    - **Currently, only affects the API path /register. Has no effect in the frontend.**
- **MTSU_JWT_SECRET**=secret123
    - Secret to sign JWTs for login sessions in the backend. Recommended to change.
- **MTSU_THUMBNAIL_FORMAT**=webp
  - Supported formats: webp 
  - AVIF support is planned. AVIF is said to take 20% longer to encode, but it compresses to 20% smaller size compared to WebP.
- **MTSU_LTR**=true
    - Set to false to use right-to-left (RTL) default for galleries. Otherwise, defaults to left-to-right (like Japanese manga).

## 📝 Mangatsu Web - Configuration

- **NEXT_PUBLIC_MANGATSU_API_URL**=https://mangatsu-api.example.com
    - URL to the backend API server.
- **NEXT_PUBLIC_INTERNAL_MANGATSU_API_URL**=http://mtsuserver:5050
    - Internal URL to the backend server. Required when running both containers locally without external network.
    - For example, if both containers are running on the same network, the value should probably be "http://mtsuserver:5050".
- **NEXT_MANGATSU_IMAGE_HOSTNAME**=mangatsu-api.example.com
    - Hostname or the domain where images are hosted. Usually the same as the domain in the API URL.
- **PORT**=3030
    - Port to run the web client on. If you change this and also use Docker Compose, remember to update the first port of frontend's ports in the `docker-compose.yml` file.


docs/CHANGELOG.md
# Changelog

All notable changes of this project will be documented in this file.

> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and this project adheres
> to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.8.0] - 2024-03-23

### Added

- Crop gallery covers to fit the aspect ratio of the thumbnail (2:3) better. Also, when it comes to multipage images, Mangatsu tries to find the cover.
- Returning the total count (TotalCount) of galleries when fetching (helps with pagination)
- meta_title_hash column to save the hash of the gallery title. Helps to save time and resources when parsing titles for metadata
- MTSU_FUZZY_SEARCH_SIMILARITY environmental variable to set the similarity threshold for fuzzy search. Default is 0.7 (70%)
- MTSU_LTR environmental variable. For now used for the default to crop gallery covers when needed. Otherwise, the language of the gallery title is used to determine it
- fuzzy (true|false) query parameter to enable or disable fuzzy search when starting metadata task

### Fixed

- Shuffling not working when fetching grouped galleries
- Paging when shuffling enabled (was returning the same first page)
- Various issues with updating galleries
- Crash when trying to count images of an archive that could not be opened
- A rare error when DB UPDATE was rollbacked
- DB lock when creating thumbnails with semaphore

### Changed

- Improve performance when fetching galleries by constructing SQL filters only once
- Return OrderedMap ({ []uuid, []{uuid: gallery} }) instead of a normal mapped results ([]{uuid: gallery}) for grouped galleries
- Store page thumbnails in thumbnails/[gallery uuid]/p/ directory instead of thumbnails/[gallery uuid]/. Cover is stored as thumbnails/[gallery uuid]/cover.[ext]
- Refactored the way galleries are fetched, updated and deleted
- Better gallery validation when updating
- Improve metadata parsing logic and performance
- Improve gallery update and fetch logic

### Docs

- Changed NEXT_INTERNAL_MANGATSU_API_URL to NEXT_PUBLIC_INTERNAL_MANGATSU_API_URL

## [0.7.5] - 2024-03-21

### Added

- MTSU_THUMBNAIL_FORMAT environmental variable. Currently only supports webp
- AVIF to supported gallery images

### Fixed

- A rare occurrence if thumbnail generation failing
- Checking file and dir path more reliably
- Fetching random galleries with a seed did not return the same result

### Changed

- Go to 1.22
- Updated dependencies
- math/rand to math/rand/v2

### Docs

- Updated README, example docker-compose.yml files and example.env

## [0.7.4] - 2024-01-23

### Fixed

- Server crash (nil pointer dereference) when trying to log in anonymously without cookie age

## [0.7.3] - 2024-01-22

### Added

- Super admin role for the initially created user

### Changed

- Allow deleting admin, but not super admin users

## [0.7.2] - 2024-01-22

### Added

- Automatically migrating bcrypt passwords to argon2
- Username, password, session name, cookie age validations

### Changed

- Bcrypt to Argon2id for hashing passwords

## [0.7.1] - 2024-01-21

### Added

- Return expiredIn value (unix time) for anonymous login as well
- On logout, clear the cookie even if the token is incorrect
- Status gone (410) to errorHandler

### Fixed

- Regression where restricted/anonymous access was not working
- A bug with returning 401 when not needed
- Setting thumbnail count when generating thumbnails

### Changed

- Updated dependencies

## [0.7.0] - 2024-01-09

### Added

- **Endpoint to return task status `/status`**
- **Query param "meta" to read the metadata without having to read the gallery itself**
- Return current session ID on login
- `deleted` and `page_thumbnails` column to gallery table
- tif (tiff) and heif image extensions to supported extensions

### Fixed

- Error message casing
- Regression where the API would return an error when a gallery was not found
-

### Changed

- Minimum Go version to 1.21
- Updated dependencies
- Enabled Profile-guided optimization ([go.dev/doc/pgo](https://go.dev/doc/pgo))
- Regenerated jet models
- Deprecated rand.Seed to rand.New
- Deprecated ioutil.ReadDir to os.ReadDir

### Docs

- Instructions to install goose and jet tools

## [0.6.1] - 2023-05-30

**Released packages also on GHCR (GitHub Container Registry) alongside DockerHub:**

- [ghcr.io/mangatsu/server](https://github.com/Mangatsu/server/pkgs/container/server)
- [ghcr.io/mangatsu/web](https://github.com/Mangatsu/server/pkgs/container/server)

### Added

- More secure logic to handle CORS
    - An env `MTSU_STRICT_ACAO` ('true' or 'false') to disable or enable it

### Changed

- Update JWT package to v5

## [0.6.0] - 2023-05-29

### Added

- `MTSU_ENV` environmental: `development` or `production`
- `MTSU_LOG_LEVEL` environmental: `debug`, `info`, `warn` or `error`

### Changed

- Replace [logrus](https://github.com/sirupsen/logrus) with [zap](https://github.com/uber-go/zap)
    - and refactor accordingly

## [0.5.1] - 2023-05-28

### Added

- New environmental: MTSU_DOMAIN
- Domain property to Cookies send back to browser
    - As, if Domain is specified, then subdomains are always included in the allowed domains.
    - https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#define_where_cookies_are_sent

### Fixed

- Internal version number

## [0.5.0] - 2023-05-28

### Added

- Support for cookies (JWT)
- Support for 7zip (7z) galleries

### Fixed

- Broken response when there were no galleries

### Changed

- Minimum Go version to 1.20
- Updated dependencies

### Docs

- Remove next-auth references
    - next-auth will be removed in Mangatsu Web v0.5.0

## [0.4.3] - 2022-04-17

### Added

- Periodically prune sessions

### Changed

- Increase maximum session duration to 1 year

## [0.4.2] - 2022-04-16

### Added

- Show internal server error messages in console
- API endpoint for gallery count `/galleries/count`
- Seed parameter for shuffling gallery results

### Fixed

- Setting a custom SQLite db filename

internal/config/config.go
package config

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/Mangatsu/server/pkg/log"
	"github.com/joho/godotenv"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

type Layout string

const (
	Freeform   Layout = "freeform"
	Structured        = "structured"
)

type Visibility string

const (
	Private    Visibility = "private"
	Restricted            = "restricted"
	Public                = "public"
)

type ImageFormat string

const (
	WebP ImageFormat = "webp"
	AVIF             = "avif"
)

type CacheOptions struct {
	WebServer bool
	TTL       time.Duration
	Size      uint64
}

type GalleryOptions struct {
	ThumbnailFormat       ImageFormat
	FuzzySearchSimilarity float64
	LTR                   bool
}

type OptionsModel struct {
	Environment    log.Environment
	Domain         string
	Hostname       string
	Port           string
	StrictACAO     bool
	Registrations  bool
	Visibility     Visibility
	DB             DBOptions
	Cache          CacheOptions
	GalleryOptions GalleryOptions
}

type CredentialsModel struct {
	JWTSecret  string
	Passphrase string
}

var AppEnvironment log.Environment
var LogLevel zapcore.Level

// Options stores the global configuration for the server
var Options *OptionsModel

// Credentials stores the JWT secret, and optionally a passphrase and credentials for the db
var Credentials *CredentialsModel // TODO: Encrypt in memory?

// LoadEnv loads the environment variables.
func LoadEnv() {
	var err = godotenv.Load()
	if err != nil {
		fmt.Println("No .env file or environmentals found.")
	}

	loadEnvironment()
	loadLogLevel()
}

// SetEnv sets the environment variables into Options and Credentials
func SetEnv() {
	Options = &OptionsModel{
		Domain:        domain(),
		Hostname:      hostname(),
		Port:          port(),
		StrictACAO:    acao(),
		Registrations: registrationsEnabled(),
		Visibility:    currentVisibility(),
		DB: DBOptions{
			Name:       dbName(),
			Migrations: dbMigrationsEnabled(),
		},
		Cache: CacheOptions{
			WebServer: cacheServerEnabled(),
			TTL:       cacheTTL(),
			Size:      cacheSize(),
		},
		GalleryOptions: GalleryOptions{
			ThumbnailFormat:       thumbnailFormat(),
			FuzzySearchSimilarity: fuzzySearchSimilarity(),
			LTR:                   defaultLTR(),
		},
	}

	Credentials = &CredentialsModel{
		JWTSecret:  jwtSecret(),
		Passphrase: restrictedPassphrase(),
	}
}

func GetInitialAdmin() (string, string) {
	username := os.Getenv("MTSU_INITIAL_ADMIN_NAME")
	password := os.Getenv("MTSU_INITIAL_ADMIN_PW")
	if username == "" {
		username = "admin"
	}
	if password == "" {
		password = "admin321"
	}
	return username, password
}

func loadEnvironment() {
	value := os.Getenv("MTSU_ENV")
	if value == "production" {
		AppEnvironment = log.Production
		return
	}
	AppEnvironment = log.Development
}

func loadLogLevel() {
	value := os.Getenv("MTSU_LOG_LEVEL")
	switch value {
	case "debug":
		LogLevel = zap.DebugLevel
	case "warn":
		LogLevel = zap.WarnLevel
	case "error":
		LogLevel = zap.ErrorLevel
	default:
		LogLevel = zap.InfoLevel
	}
}

func domain() string {
	return os.Getenv("MTSU_DOMAIN")
}

func hostname() string {
	value := os.Getenv("MTSU_HOSTNAME")
	if value == "" {
		return "localhost"
	}
	return value
}

func port() string {
	value := os.Getenv("MTSU_PORT")
	if value == "" {
		return "5050"
	}
	return value
}

func acao() bool {
	return os.Getenv("MTSU_STRICT_ACAO") == "true"
}

func cacheServerEnabled() bool {
	value := os.Getenv("MTSU_DISABLE_CACHE_SERVER")
	if value == "true" {
		return false
	}
	return true
}

func registrationsEnabled() bool {
	value := os.Getenv("MTSU_REGISTRATIONS")
	if value == "true" {
		return true
	}
	return false
}

func currentVisibility() Visibility {
	value := os.Getenv("MTSU_VISIBILITY")
	switch value {
	case "public":
		return Public
	case "restricted":
		return Restricted
	default:
		return Private
	}
}

func restrictedPassphrase() string {
	value := os.Getenv("MTSU_RESTRICTED_PASSPHRASE")
	if value == "" {
		if currentVisibility() == Restricted {
			log.Z.Warn("MTSU_RESTRICTED_PASSPHRASE is not set. Defaulting to 's3cr3t'.")
		}
		return "s3cr3t"
	}
	return value
}

func jwtSecret() string {
	value := os.Getenv("MTSU_JWT_SECRET")
	if value == "" {
		log.Z.Error("MTSU_JWT_SECRET is not set. An unsecure secret will be used instead. DO NOT USE IN PRODUCTION.")
		return "9Wag7sMvKl3aF6K5lwIg6TI42ia2f6BstZAVrdJIq8Mp38lnl7UzQMC1qjKyZCBzHFGbbqsA0gKcHqDuyXQAhWoJ0lcx4K5q"
	}
	return value
}

func cacheTTL() time.Duration {
	defaultDuration := time.Hour * 336
	minDuration := time.Minute * 15

	value := os.Getenv("MTSU_CACHE_TTL")
	if value == "" {
		return defaultDuration
	}

	duration, err := time.ParseDuration(value)
	if err != nil {
		log.Z.Error(value + " is not a valid TTL for MTSU_CACHE_TTL. Defaulting to 336h (14 days).")
		return defaultDuration
	}

	if duration < minDuration {
		log.Z.Info("Minimum TTL is 15 minutes. Defaulting to 15 minutes.")
		return minDuration
	}

	return duration
}

func cacheSize() uint64 {
	defaultSize := uint64(20000)
	minSize := uint64(100)

	value := os.Getenv("MTSU_CACHE_SIZE")
	if value == "" {
		return defaultSize
	}

	size, err := strconv.ParseUint(value, 10, 64)
	if err != nil {
		log.Z.Error(value + " is not a valid TTL for MTSU_CACHE_SIZE. Defaulting to 20 000 MiB.")
		return defaultSize
	}

	if size < minSize {
		log.Z.Warn("Minimum TTL is 100 MiB. Defaulting to 100 MiB.")
		return minSize
	}

	return size
}

func thumbnailFormat() ImageFormat {
	// TODO: Add support for AVIF
	//value := os.Getenv("MTSU_THUMBNAIL_FORMAT")
	//if value == "avif" {
	//	return AVIF
	//}
	return WebP
}

func fuzzySearchSimilarity() float64 {
	value := os.Getenv("MTSU_FUZZY_SEARCH_SIMILARITY")
	if value == "" {
		return 0.7
	}

	similarity, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Z.Error(value + " is not a valid similarity value for MTSU_FUZZY_SEARCH_SIMILARITY. Defaulting to 0.7.")
		return 0.7
	}
	if similarity < 0.1 {
		return 0.1
	}
	if similarity > 1 {
		return 1
	}
	return similarity
}

func defaultLTR() bool {
	value := os.Getenv("MTSU_LTR")
	if value == "false" {
		return false
	}
	return true
}
